guppylang/compiler/expr_compiler.py

```diff
@@ -533,6 +539,33 @@ def visit_BarrierExpr(self, node: BarrierExpr) -> Wire:
         self._update_inout_ports(node.args, iter(barrier_n), node.func_ty)
         return self._pack_returns([], NoneType())
 
+    def visit_StateResultExpr(self, node: StateResultExpr) -> Wire:
+        num_qubits = len(node.args) - 1
+        num_qubits_arg = ht.BoundedNatArg(num_qubits)
+        args = [ht.StringArg(node.tag), num_qubits_arg]
+        sig = ht.FunctionType(
+            [array_type(ht.Qubit, num_qubits_arg)],
+            [array_type(ht.Qubit, num_qubits_arg)],
+        )
+        op = ops.Custom(
+            op_name="StateResult", signature=sig, args=args, extension="tket2.debug"
+        )
+
+        qubits_in = [self.visit(e) for e in node.args[1:]]
+        # TODO: This should be caught during checking.
```

Comment:
Importing `qubit` for the `StateResultChecker` in `checker.py` leads to circular import errors - not sure if there is an obvious other place to move it to